Resolves issue #54

Platnumrcn14 opened the pull request:
Resolves issue #54. Adds PotManager.cs containing functions to allow TexasHoldemManager to interact with the pot 
-InitializePot will create the pot and adds antes to pot at the beginning of a round 
-AddToPot adds a successful round of betting to the pot assuming there are no all in bets 
-CreateSidePots will create the appropriate number of side pots based on the round of betting 
-DistributePot will distribute the appropriate payout, accounting for split winners if needed 
-GetNumPots returns the number of pots present, usable for DistributePot if multiple pots are present 
-GetPotAmounts will return the total value of the pots, usable for front end design to show pot total 
-ResetPots will clear the list of all pots, usable at the end of a round 
-Draw is a slotholder, usable for future development if needed

Platnumrcn14 committed:
Resolves issue #54
Adds PotManager.cs containing functions to allow TexasHoldemManager to interact with the pot
-InitializePot will create the pot and adds antes to pot at the beginning of a round
-AddToPot adds a successful round of betting to the pot assuming there are no all in bets
-CreateSidePots will create the appropriate number of side pots based on the round of betting
-DistributePot will distribute the appropriate payout, accounting for split winners if needed
-GetNumPots returns the number of pots present, usable for DistributePot if multiple pots are present
-GetPotAmounts will return the total value of the pots, usable for front end design to show pot total
-ResetPots will clear the list of all pots, usable at the end of a round
-Draw is a slotholder, usable for future development if needed

jacobwilki01 commented on CardsCashCasino/Manager/PotManager.cs:
```diff
@@ -0,0 +1,145 @@
+        /// </summary>
+        private int _potPointer;
+        
+
```
remove whitespace

Platnumrcn14 replied:
Easy peasy.

jacobwilki01 commented on CardsCashCasino/Manager/PotManager.cs:
```diff
@@ -0,0 +1,145 @@
+    /// <summary>
+    /// Manages the pot in a Texas Hold'em game, including adding and distributing chips.
+    /// </summary>
+    public class PotManager
```
I want to rework this a bit. Not really a fan of a bunch of floating ints and pointers. Instead, lets make a Pot class that stores this info, and a list of those objects in HoldEmManager.

Platnumrcn14 replied:
Got it. Now implemented in HoldEmManager.

Platnumrcn14 committed:
Merge from develop

Platnumrcn14 committed:
-InitializePot will create the pot and adds antes to pot at the beginning of a round
-AddToPot adds a successful round of betting to the pot assuming there are no all in bets
-CreateSidePots will create the appropriate number of side pots based on the round of betting if all-in bet(s) are present
-DistributePot distributes the amounts contained in the pot to the appropriate winner
-GetNumPots returns the number of pots present, usable for DistributePot if multiple pots are present
-GetPotAmounts will return the total value of the pots, usable for front end design to show pot total
-ResetPots will clear the list of all pots, usable at the end of a round

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+            // List to hold the pots for the game
+            public List<Pot> Pots;
+
+            public int Total { get; set; }
```
remove these instances of attributes, so they are only handled through the Pot.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
             timer.Stop();
             timer.Dispose();
         }
+        public class PotManager
```
lets name this something slightly different so its more descriptive `TexasHoldEmPotManager` is better.

Also, lets not nest this within HoldEmManager. Have it sit within the namespace, not the class. Otherwise, it should be `internal` not `public`.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+            /// <summary>
+            /// Manages the pot in a Texas Hold'em game, including adding and distributing chips.
+            /// </summary>
+            public class Pot
```
Let's move this so it is not nested within PotManager and rename it to `TexasHoldEmPot`. Just make it nested within TexasHoldEmManager.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+            /// </summary>
+            public class Pot
+            {
+                public int Total { get; set; }
```
These attributes all need summary comments..

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+                public PotType PotType { get; set; }
+                public List<int> EligiblePlayers { get; set; }
 
+                public Pot(PotType type, int total, List<int> eligiblePlayers)
```
Make the only constructor parameter here the type.

Since total is always initialized at zero, you can just edit the line above to be `public int Total { get; set; } = 0;`.

I am not a fan of pushing a list of eligible players into the constructor, it gets really messy with C# OR just modifying the total value directly. Instead, let's create a few helper methods within here that are descriptive:
---`IncrementPot(int value)`
---`DecrementPot(int value)`
---`AddEligiblePlayer(int player)`
---`RemoveEligiblePlayer(int player)`.

Platnumrcn14 replied:
Added IncrementPot, DecrementPot, and RemoveEligiblePlayer. AddEligiblePlayer shouldn't be necessary since players wouldn't be gaining further eligibility to different pots, players only lose eligibility when folding or going all in. Currently I configured InitiatePot to add all players to the EligiblePlayers list for the MAIN pot, players can be removed from additional pots as needed with the RemoveEligiblePlayer function suggested and implemented.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+                    EligiblePlayers = eligiblePlayers;
+                }
+            }
+            /// <summary>
```
make sure there is a space between the summary comment and the method before. this is true of all of the methods here.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+                {
+                    return;
+                }
+                else
```
dont need an else here.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+            /// <param name="playersBets"> list of bets that were placed but folded due to a raise in that round</param>
+            public void AddFoldedBetsToPot(List<int> playerBets)
+            {
+                if (playerBets.Count() == 0)
```
use the Linq method `if (playerBets.Any())` instead.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+            {
+                Pots.Add(new Pot(PotType.MAIN, 0, playersActive));
+
+                if (playerBets.All(n => n == playerBets[0])) //if all bets match the ante
```
Let's make `n` a more descriptive variable. So `bet` or something.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+            /// Adds a bets to the pot.
+            /// </summary>
+            /// <param name="_currentBet">The amount each player is contributing to the pot.</param>
+            /// <param name="numPlayers">The number of players contributing to the pot.</param>
```
make sure the summary comments remain active. Currently, this is referencing a parameter that is no longer there (`numPlayers`).

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+            /// <param name="_currentBet">The amount to add to the pot.</param>
+            /// <param name="playerBets">The list of bets from players still active after the round of betting.</param>
+            /// <param name="playersActive">The list of players eligible to win the main pot, all in players will be subtracted for side pot</param>
+            public void CreateSidePots(int _currentBet, List<int> playerBets, List<int> playersActive)
```
```suggestion
            public void CreateSidePots(int currentBet, List<int> playerBets, List<int> playersActive)
```

remove the underscore in the parameter name.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+            public void CreateSidePots(int _currentBet, List<int> playerBets, List<int> playersActive)
+            {
+                int allInBet = 0; //initiating variable that will hold all in wager
+                int numBets = playerBets.Count();
```
`Count` is not a method, its an attribute. You just need to do `.Count`.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+                        AddToPot(allInBet, Enumerable.Repeat(allInBet, playerBets.Count()).ToList(), playersActive);
+
+                        //subtracting all-in bet value from all other bets, the all-in bet value has already been added to a pot
+                        for (int player = 0; player < playerBets.Count(); player++)
```
```suggestion
                        for (int player = 0; player < numBets; player++)
```

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+                for (int bets = 0; bets < numBets; bets++)
+                {
+                    //at least 1 side pot is needed if function is called, lowest bet will be an all-in bet.
+                    if (bets == 0)
```
instead of nesting this as an if clause within the for loop, lets just call all of this logic once before the for loop and unnest it a bit.

Platnumrcn14 replied:
Removed the for loop as requested. Let me know if there are any additional changes I can implement for it!

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+                        {
+                            playerBets[player] -= allInBet; // Subtract the amount put into pot from each bet
+                        }
+                        _currentBet -= allInBet;
```
does this edit need to be reflected outside of this method? if so, you need to pass it in as a reference parameter using the `ref` keyword.

Platnumrcn14 replied:
It shouldn't need to be reflected outside of the method, it's purpose will be fully served within the CreateSidePots method for splitting what portion of each bet belongs in the MAIN pot and the SIDE pot

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+                        Pots[0].EligiblePlayers = playersActive;
+                        Pots[0].Total = 0; //reseting active pot value to empty
+                    }
+                    else if (playerBets.Count() == 1)
```
this clause will always be hit, and will break the loop meaning the rest of the code will never be hit. Did you mean to put a `continue` in this if statement?

Platnumrcn14 replied:
In the refactored code it should still be that else if statement, it is meant to terminate the method once encountered. It would be encountered in the scenario where only one person has funds remaining to bet in the round.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+            /// <param name="winners">Number of players that have won the pot.</param>
+            public int DistributePot(int winners)
+            {
+                int _payout = Pots.Last().Total / winners; //splits payout if more than 1 winner is present
```
we need to error check to make sure winners is not zero. you also need to account for rounding here, because it may round up more money than exists.

Platnumrcn14 replied:
Added an error check for 0. To my knowledge the only operation that should occur is an integer by integer division, it should round down by default.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -804,9 +810,198 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+                Pots.RemoveAt(Pots.Count - 1);
+                return _payout;
+            }
+            public List<int> GetPotAmounts()
```
summary comment. this is an attribute, so it should be at the top of the class, not within the methods.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -47,6 +47,11 @@ public enum BettingPhase
         TURN,
         RIVER
     }
+    public enum PotType
```
summary comment.

jacobwilki01 reviewed (changes requested):
This is a lot of logic being added. I'd like you to test this using unit tests.

Platnumrcn14 committed:
-Corrected changes requested
-Refactored CreateSidePots method to exclude for loop, instead calling itself recursively if needed
-Added IncrementPot, DecrementPot, RemoveEligiblePlayer, and RemoveFoldedPlayer methods to more easily account for pot funds and player-pot eligability
-Created Unit tests for each scenerio that PotManager can encounter.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -738,7 +741,7 @@ private void HandlePlayerAction(int playerIndex)
         }
-        
+
         private PokerAction GetUserAction()
```
still missing summary comments

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -751,15 +754,15 @@ private PokerAction GetUserAction()
         }
-        
+
         private PokerAction GetAIAction(int playerIndex)
```
still missing summary comments
the wall of comments within this method should be within the summary comments

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -810,199 +813,249 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+        /// <summary>
+        /// Manages the pot in a Texas Hold'em game, including adding and distributing chips.
+        /// </summary>
+        public class TexasHoldEmPot
```
this should not be nested.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -810,199 +813,249 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+
+        public TexasHoldEmPotManager()
+        {
+            Pots = new List<TexasHoldEmManager.TexasHoldEmPot>(); // Initialize the Pots list
```
this does not need to be initalized in the constructor. just initialize it where it is created above, and remove the constructor.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -810,199 +813,249 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+    public class TexasHoldEmPotManager
+    {
+        // List to hold the pots for the game
+        public List<TexasHoldEmManager.TexasHoldEmPot> Pots { get; set; }
```
summary comment!!!

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -810,199 +813,249 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+            //if all players but one are all-in 
+            else if (numBets == 1)
+            {
+                Pots.RemoveAt(0); //remove redundant pot creation
```
uh why is this happening? we should remove the code that makes this redundant pot, instead of just removing it.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -810,199 +813,249 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+        {
+            if (winners == 0)
+            {
+                throw new ArgumentException("Number of winners cannot be zero.");
```
under no circumstance should we throw an exception like this. instead, on the call to DistributePot, just confirm winners > 0 before calling and if not, just skip it because its errored out.

Throwing exceptions like this is scary, and can break the game if its something simple.

Platnumrcn14 committed:
-Updated & modified summary comments
-Repositioned TexasHoldEmPot class out of nested position, modifying usages appropriately
-Refactored code for scenerio where only one player is not all in at the end of side pot creation. Redundant pot no longer generated then removed
-Modified DistributePots to prevent division by 0, with ArgumentException removed
-Created some front end options for future implementation creating the capability for textures to be implemented for scenerios when options like check, raise, and call are unavaliable

Platnumrcn14 committed:
-Missed updating LoadContent TODOs for updated button capabilities

Platnumrcn14 committed:
-corrected additional case of throwing exception found in DecrementPot

modmo22 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -799,31 +880,313 @@ private static void OnTimeoutEvent(object source, ElapsedEventArgs e)
+    /// <summary>
+    /// Manages the pot in a Texas Hold'em game, including adding and distributing chips.
+    /// </summary>
+    public class TexasHoldEmPot
```
Is the idea that we incorporate a pot in each game? We'd copy and paste this code into each game's manager? Wouldn't it be easier to separate this into it's own module and import it into each game? It's the same question I had for the player manager in the other PR.

Other than that, this looks good

eBerkley replied:
Imo since this file is already >800 lines atp there's no reason not to at least move it to it's own file

eBerkley replied:
Gonna resolve though for the sake of time

modmo22 commented on UnitTests/HoldEmTests.cs:
```diff
@@ -1,13 +1,315 @@
 using System.Threading.Tasks;
+using static CardsCashCasino.Manager.TexasHoldEmManager;
 
 namespace CardsCashCasino.NUnit
```
These tests look spectacular. Way to go, dude. You really knocked these out of the park

eBerkley reviewed (approved):
Looks good, great work!
